src/server/transaction/server/driver.go

```diff
@@ -0,0 +1,282 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/transaction"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/serviceenv"
+	"github.com/pachyderm/pachyderm/src/server/pkg/transactiondb"
+	txnenv "github.com/pachyderm/pachyderm/src/server/pkg/transactionenv"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+)
+
+type driver struct {
+	// txnEnv stores references to other pachyderm APIServer instances so we can
+	// make calls within the same transaction without serializing through RPCs
+	txnEnv *txnenv.TransactionEnv
+
+	// etcdClient and prefix write repo and other metadata to etcd
+	etcdClient *etcd.Client
+	prefix     string
+
+	// collections
+	transactions col.Collection
+}
+
+func newDriver(
+	env *serviceenv.ServiceEnv,
+	txnEnv *txnenv.TransactionEnv,
+	etcdPrefix string,
+) (*driver, error) {
+	etcdClient := env.GetEtcdClient()
+	d := &driver{
+		txnEnv:       txnEnv,
+		etcdClient:   etcdClient,
+		prefix:       etcdPrefix,
+		transactions: transactiondb.Transactions(etcdClient, etcdPrefix),
+	}
+	return d, nil
+}
+
+func now() *types.Timestamp {
+	t, err := types.TimestampProto(time.Now())
+	if err != nil {
+		panic(err)
```

Comment:
Let's not panic here, it's unlikely to ever get hit but it makes me a bit nervous to think we could put the server in a crash loop with this, we can just swallow the error instead and return a zero time.

Reply:
Fixed here and in PFS where this was copied from.